src/console_gui.cpp

```diff
@@ -258,22 +295,34 @@ struct IConsoleWindow : Window
 				IConsoleCmdExec("clear");
 				break;
 
-			default:
-				if (_iconsole_cmdline.HandleKeyPress(key, keycode) != HKPR_NOT_HANDLED) {
+			case WKC_TAB:
+				if (_iconsole_tab_completion.AutoComplete()) {
+					this->SetDirty();
+				}
+				break;
+
+			default: {
+				HandleKeyPressResult handle_result = _iconsole_cmdline.HandleKeyPress(key, keycode);
+				if (handle_result != HKPR_NOT_HANDLED) {
+					if (handle_result == HKPR_EDITING) {
+						_iconsole_tab_completion.Reset();
```

Comment:
The current autocompletion is stateful - it cycles through suggestions in a loop on Tab.

Every time the user edits _iconsole_cmdline text buffer, we need to reset the state of autocompletion.

I have added a Reset every time we change _iconsole_cmdline in this file.

However, it is also exposed through GetFocusedTextbuf() method of Window and I don't have any way to "catch" edit events if textbuf is updated via this method.

In network_chat_gui.cpp it is fine, because we have OnEditboxChanged, but this event is not called when we type text in the console.

However, when. I tried adding OnEditboxChanged to IConsoleWindow, it was not called when typing in console and I am not sure why.

So I am not sure that this change is actually fully correct.
But manual testing didn't show any problems.

Reply:
IConsoleWindow acts like an editbox, but it's not an editbox and it doesn't have an editbox widget either. This is why OnEditboxChanged() is never called for it.

Reply:
How is GetFocusedTextbuf() works? Maybe we can remove this method, so _iconsole_cmdline will become fully private?

Reply:
`GetFocusedTextbuf()` is used only by macos, in ObjC++ functions related to text handling (selection, insertion, ...).
Console being a kind of editbox, this function can't be removed.

Reply:
I checked that actually GetFocusedTextbuf() returns a const pointer to Textbuf, so it is only used to read the current text, not modify it, so it is safe.

With this I am resolving this thread